src/c/libp2p_c.cpp

```diff
@@ -0,0 +1,584 @@
+    auto protocol_handler = std::make_shared<GenericProtocolHandler>(
+        protocol_id, handler, user_data, host);
+
+    // Register the protocol with the host - now passing the protocol handler directly
+    auto result = host->host->listenProtocol(protocol_id, protocol_handler);
+    if (!result) {
+      return LIBP2P_ERROR_PROTOCOL_ERROR;
+    }
+
+    return LIBP2P_SUCCESS;
+  } catch (...) {
+    return LIBP2P_ERROR_UNKNOWN;
+  }
+}
+
+// Stream operations implementation
+libp2p_error_t libp2p_stream_read(libp2p_stream_t *stream,
+                                  uint8_t *buffer,
+                                  size_t buffer_size,
+                                  libp2p_read_callback_t callback,
+                                  void *user_data) {
+  if (!stream || !stream->stream || !buffer || !callback) {
+    return LIBP2P_ERROR_INVALID_ARGUMENT;
+  }
+
+  try {
+    libp2p::coroSpawn(
+        *stream->host->context->io_context,
+        [stream, buffer, buffer_size, callback, user_data]()
+            -> libp2p::Coro<void> {
+          std::span<uint8_t> buffer_span{buffer, buffer_size};
+          auto n_res = co_await stream->stream->readSome(buffer_span);
+          if (n_res.has_value()) {
+            callback(stream, buffer, n_res.value(), LIBP2P_SUCCESS, user_data);
+          } else {
+            callback(
+                stream, nullptr, 0, convert_error(n_res.error()), user_data);
+          }
+        });
+    return LIBP2P_SUCCESS;
+  } catch (...) {
+    return LIBP2P_ERROR_UNKNOWN;
+  }
+}
+
+libp2p_error_t libp2p_stream_write(libp2p_stream_t *stream,
+                                   const uint8_t *data,
+                                   size_t size,
+                                   libp2p_write_callback_t callback,
+                                   void *user_data) {
+  if (!stream || !stream->stream || !data || !callback) {
+    return LIBP2P_ERROR_INVALID_ARGUMENT;
+  }
+
+  try {
+    libp2p::coroSpawn(
+        *stream->host->context->io_context,
+        [stream, data, size, callback, user_data]() -> libp2p::Coro<void> {
+          std::span<const uint8_t> data_span{data, size};
+          auto result = co_await libp2p::write(stream->stream, data_span);
+          if (result.has_value()) {
+            callback(stream, size, LIBP2P_SUCCESS, user_data);
+          } else {
+            callback(stream, 0, convert_error(result.error()), user_data);
+          }
+        });
+    return LIBP2P_SUCCESS;
+  } catch (...) {
+    return LIBP2P_ERROR_UNKNOWN;
+  }
+}
+
+libp2p_error_t libp2p_stream_close(libp2p_stream_t *stream) {
+  if (!stream || !stream->stream) {
+    return LIBP2P_ERROR_INVALID_ARGUMENT;
+  }
+
+  try {
+    if (auto result = stream->stream->close(); !result) {
+      return convert_error(result.error());
+    }
+    return LIBP2P_SUCCESS;
+  } catch (...) {
+    return LIBP2P_ERROR_UNKNOWN;
+  }
+}
+
+bool libp2p_stream_is_closed(libp2p_stream_t *stream) {
```

Comment:
`isClosed` is ambiguous,
because it can `isClosed=false` doesn't guarantee that read/write will succeed.
Caller of read/write will receive error and mark stream as closed (e.g. stop reading, remove from active stream list, ...).
```suggestion
```